linera-witty/src/runtime/wasmer/parameters.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Representation of Wasmer function parameter types.
+
+use crate::{memory_layout::FlatLayout, primitive_types::FlatType};
+use frunk::{hlist, hlist_pat, HList};
+use wasmer::FromToNativeWasmType;
+
+/// Conversions between flat layouts and Wasmer parameter types.
+pub trait WasmerParameters: FlatLayout {
+    /// The type Wasmer uses to represent the parameters in a function imported from a guest.
+    type ImportParameters;
+
+    /// The type Wasmer uses to represent the parameters in a function exported from a host.
+    type ExportParameters;
+
+    /// Converts from this flat layout into Wasmer's representation for functions imported from a
+    /// guest.
+    fn into_wasmer(self) -> Self::ImportParameters;
+
+    /// Converts from Wasmer's representation for functions exported from the host into this flat
+    /// layout.
+    fn from_wasmer(parameters: Self::ExportParameters) -> Self;
+}
+
+impl WasmerParameters for HList![] {
+    type ImportParameters = ();
+    type ExportParameters = ();
+
+    fn into_wasmer(self) -> Self::ImportParameters {}
+
+    fn from_wasmer((): Self::ExportParameters) -> Self {
+        hlist![]
+    }
+}
+
+impl<Parameter> WasmerParameters for HList![Parameter]
+where
+    Parameter: FlatType + FromToNativeWasmType,
+{
+    type ImportParameters = Parameter;
+    type ExportParameters = (Parameter,);
+
+    #[allow(clippy::unused_unit)]
+    fn into_wasmer(self) -> Self::ImportParameters {
+        let hlist_pat![parameter] = self;
+
+        parameter
+    }
+
+    fn from_wasmer((parameter,): Self::ExportParameters) -> Self {
+        hlist![parameter]
+    }
+}
+
+/// Helper macro to implement [`WasmerParameters`] for flat layouts up to the maximum limit.
+///
+/// The maximum number of parameters is defined by the [canonical
+/// ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md#flattening)
+/// as the `MAX_FLAT_PARAMS` constant.
```

Comment:
The value 16 is hardcoded in this library in various places, right? Maybe explain this in the comment.

Reply:
Done. I tried to improve the related documentation and mentioned the `MAX_FLAT_*` constants so it's easier to `grep` for them.